ITE: soc/power: add power management

GTLin08 opened the pull request:
Add the power state of deep doze. When system enters deep doze, the
clock of CPU and EC can be stopped to reduce power consumption. And
enable the UART Rx WUI before entering deep doze to wake up EC and
CPU.

Tested on it8xxx2_evb board. It will reduce 5.25mA when system enters
deep doze mode.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504)
<!-- Reviewable:end -->

GTLin08 committed:
ITE: soc/power: add power management

Add the power state of deep doze. When system enters deep doze, the
clock of CPU and EC can be stopped to reduce power consumption. And
enable the UART Rx WUI before entering deep doze to wake up EC and
CPU.

Tested on it8xxx2_evb board. It will reduce 5.25mA when system enters
deep doze mode.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

keith-zephyr reviewed (changes requested):
> Reviewed 5 of 7 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504)* status: all files reviewed, 5 unresolved discussions (waiting on @brockus-zephyr, @GTLin08, @henrikbrixandersen, @jackrosenthal, @katsuster, @kgugala, @mgielda, @nashif, @pgielda, @sjg20, @tejlmand, @tgorochowik, and @yperess)

___
*[boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig, line 39 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmT_5wJD9du0zRzMLYg:-MmT_5wJD9du0zRzMLYh:b-yu1i5s) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/boards/riscv/it8xxx2_evb/it8xxx2_evb_defconfig#L39)):*
> ```
> CONFIG_PM=n
> CONFIG_PM_POLICY_APP=n
> ```
These are the defaults, correct?  I think this can be deleted.
___
*[dts/riscv/it8xxx2.dtsi, line 35 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmTaSXzCvAYuE1uy-d5:-MmTaSXzCvAYuE1uy-d6:bst35pw) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/dts/riscv/it8xxx2.dtsi#L35)):*
> ```
> 			compatible = "zephyr,power-state";
> 			power-state-name = "standby";
> 			substate-id = <1>;
> ```

`substate-id` isn't used by ite/common/power.c, so this can be deleted.
___
*[soc/riscv/riscv-ite/common/power.c, line 68 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmTg3OW0Ns8NqzaCIKA:-MmTg3OW0Ns8NqzaCIKB:b-eljuf4) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/soc/riscv/riscv-ite/common/power.c#L68)):*
<details><summary><i alt="summary">Quoted 12 lines of code…</i></summary>

> ```
> #if DT_NODE_HAS_STATUS(DT_NODELABEL(uart1), okay)
> 	/* Enable UART1 WUI */
> 	ret = gpio_pin_interrupt_configure(DEVICE_DT_GET(DT_NODELABEL(gpiob)),
> 					   0, GPIO_INT_TRIG_LOW);
> 	if (ret) {
> 		LOG_ERR("Failed to configure UART1 WUI (ret %d)", ret);
> 		return;
> 	}
> 	/* If nothing remains to be transmitted. */
> 	while (!uart_irq_tx_complete(uart1_dev))
> 		;
> #endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart1), okay) */
> ```
</details>

Normally this code should go into the device driver by defining a `pm_action_cb` function in the `DEVICE_DT_INST_DEFINE` macro.  However, I see the IT8xxx2 uses the shared ns16550.c driver, which does not provide a power management callback.

You could create driver to handle IT8xxx2 specific UART features.  For instance:
``` 
  uart1: uart@f02700 {
    compatible = "ns16550";
    /* ... */
  }
  ite_uart1_wrapper: {
    compatible = "uart_it8xxx2";
    /* ... */
  }
```
The uart_it8xxx2 driver could then define the `pm_action_cb` and this code could be moved out of this file.  You can experiment with this in a separate PR.
___
*[soc/riscv/riscv-ite/common/power_policy.c, line 12 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmTm-86AUdRbA2D6Lxz:-MmTm-86AUdRbA2D6Ly-:bctbm7i) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/soc/riscv/riscv-ite/common/power_policy.c#L12)):*
> ```C
> #include <pm/policy.h>
> 
> __weak struct pm_state_info pm_policy_next_state(int32_t ticks)
> ```

This is supposed to be defined by the application.  I don't think this belongs in the soc support.
___
*[soc/riscv/riscv-ite/it8xxx2/soc.c, line 144 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmTmHHMDoaeU8Tqk9pc:-MmTmHHMDoaeU8Tqk9pd:b7f8u99) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/soc/riscv/riscv-ite/it8xxx2/soc.c#L144)):*
> ```C
> extern volatile int wait_interrupt_fired;
> 
> ALWAYS_INLINE void riscv_idle(enum chip_pll_mode mode, unsigned int key)
> ```

 ALWAYS_INLINE is only valid for static functions.  If you need this to be inline, you will have to move the function body into the header file.


<!-- Sent from Reviewable.io -->

GTLin08 commented:
> 
> > ```
> > CONFIG_PM=n
> > CONFIG_PM_POLICY_APP=n
> > ```
> 
> These are the defaults, correct? I think this can be deleted.

Yes, it's default. Deleted. 

> _[dts/riscv/it8xxx2.dtsi, line 35 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmTaSXzCvAYuE1uy-d5:-MmTaSXzCvAYuE1uy-d6:bst35pw) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/dts/riscv/it8xxx2.dtsi#L35)):_
> 
> > ```
> > 			compatible = "zephyr,power-state";
> > 			power-state-name = "standby";
> > 			substate-id = <1>;
> > ```
> 
> `substate-id` isn't used by ite/common/power.c, so this can be deleted.

Done

> _[soc/riscv/riscv-ite/common/power.c, line 68 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmTg3OW0Ns8NqzaCIKA:-MmTg3OW0Ns8NqzaCIKB:b-eljuf4) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/soc/riscv/riscv-ite/common/power.c#L68)):_
> 
> Quoted 12 lines of code…
> Normally this code should go into the device driver by defining a `pm_action_cb` function in the `DEVICE_DT_INST_DEFINE` macro. However, I see the IT8xxx2 uses the shared ns16550.c driver, which does not provide a power management callback.
> 
> You could create driver to handle IT8xxx2 specific UART features. For instance:
> 
> ```
>   uart1: uart@f02700 {
>     compatible = "ns16550";
>     /* ... */
>   }
>   ite_uart1_wrapper: {
>     compatible = "uart_it8xxx2";
>     /* ... */
>   }
> ```
> 
> The uart_it8xxx2 driver could then define the `pm_action_cb` and this code could be moved out of this file. You can experiment with this in a separate PR.

 Thank you for your information, It is currently under development and will be in another PR.

> _[soc/riscv/riscv-ite/common/power_policy.c, line 12 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmTm-86AUdRbA2D6Lxz:-MmTm-86AUdRbA2D6Ly-:bctbm7i) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/soc/riscv/riscv-ite/common/power_policy.c#L12)):_
> 
> > ```c
> > #include <pm/policy.h>
> > 
> > __weak struct pm_state_info pm_policy_next_state(int32_t ticks)
> > ```
> 
> This is supposed to be defined by the application. I don't think this belongs in the soc support.

It's removed. 

> _[soc/riscv/riscv-ite/it8xxx2/soc.c, line 144 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-MmTmHHMDoaeU8Tqk9pc:-MmTmHHMDoaeU8Tqk9pd:b7f8u99) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/381960da9c0ecac117de11a4b6bb6c501028d736/soc/riscv/riscv-ite/it8xxx2/soc.c#L144)):_
> 
> > ```c
> > extern volatile int wait_interrupt_fired;
> > 
> > ALWAYS_INLINE void riscv_idle(enum chip_pll_mode mode, unsigned int key)
> > ```
> 
> ALWAYS_INLINE is only valid for static functions. If you need this to be inline, you will have to move the function body into the header file.

Ok, deleted it.

keith-zephyr reviewed (approved):
> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504)* status: all files reviewed, 6 unresolved discussions (waiting on @brockus-zephyr, @ceolin, @GTLin08, @henrikbrixandersen, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @mgielda, @nashif, @pgielda, @sjg20, @tejlmand, @tgorochowik, and @yperess)



<!-- Sent from Reviewable.io -->

brockus-zephyr reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/39504#-:-MmsQXTd6X7tCZyE_F7o:bnfp4nl)* status: all files reviewed, 6 unresolved discussions (waiting on @ceolin, @GTLin08, @henrikbrixandersen, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @mgielda, @nashif, @pgielda, @sjg20, @tejlmand, @tgorochowik, and @yperess)



<!-- Sent from Reviewable.io -->
